invokeai/frontend/web/docs/WORKFLOWS_DESIGN_IMPLEMENTATION.md

```diff
@@ -0,0 +1,350 @@
+- `workflow.ts`: workflows and constituents
+
+We customize the OpenAPI schema to include additional properties on invocation and field schemas. To facilitate parsing this schema into templates, we modify/wrap the types from [openapi-types] in `openapi.ts`.
+
+### OpenAPI Schema Parsing
+
+The entrypoint for OpenAPI schema parsing is [parseSchema.ts].
+
+General logic flow:
+
+- Iterate over all invocation schema objects
+  - Extract relevant invocation-level attributes (e.g. title, type, version, etc)
+  - Iterate over the invocation's input fields
+    - [Parse each field's type](#parsing-field-types)
+    - [Build a field input template](#building-field-input-templates) from the type - either a stateful template or "generic" stateless template
+  - Iterate over the invocation's output fields
+    - Parse the field's type (same as inputs)
+    - [Build a field output template](#building-field-output-templates)
+  - Assemble the attributes and fields into an invocation template
+
+Most of these involve very straightforward `reduce`s, but the less intuitive steps are detailed below.
+
+#### Parsing Field Types
+
+Field types are represented as structured objects:
+
+```ts
+type FieldType = {
+  name: string;
+  isCollection: boolean;
+  isPolymorphic: boolean;
+};
+```
+
+The parsing logic is in `parseFieldType.ts`.
+
+There are 4 general cases for field type parsing.
+
+##### Primitive Types
+
+When a field is annotated as a primitive values (e.g. `int`, `str`, `float`), the field type parsing is fairly straightforward. The field is represented by a simple OpenAPI **schema object**, which has a `type` property.
+
+We create a field type name from this `type` string (e.g. `string` -> `StringField`).
+
+##### Complex Types
+
+When a field is annotated as a pydantic model (e.g. `ImageField`, `MainModelField`, `ControlField`), it is represented as a **reference object**. Reference objects are pointers to another schema or reference object within the schema.
+
+We need to **dereference** the schema to pull these out. Dereferencing may require recursion. We use the reference object's name directly for the field type name.
+
+> Unfortunately, at this time, we've had limited success using external libraries to deference at runtime, so we do this ourselves.
+
+##### Collection Types
+
+When a field is annotated as a list of a single type, the schema object has an `items` property. They may be a schema object or reference object and must be parsed to determine the item type.
+
+We use the item type for field type name, adding `isCollection: true` to the field type.
+
+##### Polymorphic Types
+
+When a field is annotated as a union of a type and list of that type, the schema object has an `anyOf` property, which holds a list of valid types for the union.
+
+After verifying that the union has two members (a type and list of the same type), we use the type for field type name, adding `isPolymorphic: true` to the field type.
```

Comment:
At some point, I think we could support general unions. Inputs could have an `acceptsTypes: FieldType[]` property instead of a single `type` property. 

On the frontend, the workflow migration logic I've implemented would support a change like this. On the backend, it shouldn't be too hard to update the edge validation logic.

Now is the best time to change the term "polymorphic" to something more accurate, but I'm not sure what. Maybe "hybrid"?

Reply:
Renamed to `CollectionOrScalar`